### PR TITLE
RL4J - Added TransformProcess (part 1)

### PR DESCRIPTION
**Pull request by aboulang2002**
## What changes were proposed in this pull request?

We now need some way to create the observations. This PR is the part 1 of 2 (in order to stay below 1000 lines) that will introduce something similar to DataVec's `TransformProcess`. RL4J's `TransformProcess` is defined as a sequence of steps, executed in order, that will transform the environment's raw data into an Observation instance. Currently, only one data channel is supported, but later, an environment will be able to 'publish' several data channels, for example a video feed and two audio ones.
The first version of TransformProcess has 3 types of steps:
  1) Filter: Used to skip observations
  2) Transform: This will apply a transform (e.g. RGB to grayscale) and / or convert the type of the channel (e.g. ImageWritable to INDArray)
  3) PreProcess: To apply a DataSetPreProcessor on the channel

The only requirement is that the output of the TransformProcess is:
  a) A skipped observation
or
  b) Every channels are either an instance of INDArray or DataSet.
  
The TransformProcess will be defined like this:
```
            transformProcess = TransformProcess.builder()
                    .filter(new UniformSkippingFilter(skipFrame))
                    .transform("data", new EncodableToImageWriteableTransform(height, width, 3))
                    .transform("data", new MultiImageTransform(
                            new ResizeImageTransform(rescaledWidth, rescaledHeight),
                            new ColorConversionTransform(COLOR_BGR2GRAY),
                            new CropImageTransform(offsetY, offsetX, finalHeight, finalWidth)
                    ))
                    .transform("data", new ImageWriteableToINDArrayTransform(finalHeight, finalWidth))
                    .transform("data", TemporalMergeTransform.builder()
                            .isFirstDimenstionBatch(true)
                            .build())
                    .build("data");
```

## How was this patch tested?

manual and unit tests

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X] Created tests for any significant new code additions.
- [X] Relevant tests for your changes are passing.

### CODE REVIEW

**Comment by aboulang2002** [comment]
Summary of changes:
  - Added the classes in `org.deeplearning4j.rl4j.observation.transform`
  - Added interface `Operation<TIn, TOut>` in datavec-api; `ImageTransform` and `ColumnOp` now implements this interface
  - Deleted everything in `org.deeplearning4j.rl4j.observation.preprocessor`
  - Added `datavec-api` as a dependency of `rl4j-core`
  - Added unit tests.

**Comment by aboulang2002** [comment]
Last commit:
I renamed TemporalMergeTransform to HistoryMergeTransform. After digging through Sutton's book, I think HistoryMergeTransform is a better name.
Naming things is hard.

**Comment by RobAltena** [comment]
I wonder if there is a way to do this with RL4J using DataVec but without changing DataVec.

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/helper/INDArrayHelper.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        if (shape[0] == 1) {
+            return source;
+        }
+
+        long[] newShape = new long[shape.length + 1];
+        newShape[0] = 1;
+        System.arraycopy(shape, 0, newShape, 1, shape.length);
+
+        return source.reshape(newShape);
```
Comment (treo) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/helper/INDArrayHelper.java]: Why not use [Nd4j.expandDims](https://deeplearning4j.org/api/latest/org/nd4j/linalg/factory/Nd4j.html#expandDims-org.nd4j.linalg.api.ndarray.INDArray-int-)?

Comment (aboulang2002) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/helper/INDArrayHelper.java]: Did the change, thanks!

**Comment by aboulang2002** [comment]
@RobAltena 
We could add a few wrappers in RL4J and not modify DataVec. In this PR, I only added `Operation<>` to org.datavec.image because it's the minimum I needed. But I liked the fact that any DataVec's operations could be used directly in RL4J.
Do you see an issue with adding `Operation<>` to DataVec?

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/helper/INDArrayHelper.java [hunk 2]
```diff
@@ -1,30 +1,39 @@
...
+
+/**
+ * INDArray helper methods used by RL4J
+ *
+ * @author Alexandre Boulanger
+ */
+public class INDArrayHelper {
+    /**
+     * MultiLayerNetwork and ComputationGraph expect the first dimension to be the number of examples in the INDArray.
+     * In the case of RL4J, it must be 1. This method will return a INDArray with the correct shape.
```
Comment (saudet) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/helper/INDArrayHelper.java:29]: That's not always true. There's probably something wrong with the code that needs this, but not a big deal to leave it like that for now...

**Comment by saudet** [comment]
@AlexDBlack What do you think about the small modifications to DataVec? Are you OK with that?

File: datavec/datavec-api/src/main/java/org/datavec/api/transform/Operation.java [hunk 3]
```diff
@@ -1,28 +1,20 @@
...
-/**
- * A base class for all DataSetPreProcessor that must be reset between each MDP sessions (games).
- *
- * @author Alexandre Boulanger
- */
-public abstract class ResettableDataSetPreProcessor implements DataSetPreProcessor {
-    public abstract void reset();
-}
+/*******************************************************************************
+ * Copyright (c) 2015-2020 Skymind, Inc.
```
Comment (AlexDBlack) [inline datavec/datavec-api/src/main/java/org/datavec/api/transform/Operation.java]: Let's make this:
```
Copyright (c) 2020 Konduit K.K.
```

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/observation/transform/FilterOperation.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+/*******************************************************************************
+ * Copyright (c) 2015-2019 Skymind, Inc.
```
Comment (AlexDBlack) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/observation/transform/FilterOperation.java]: New classes (not moved from elsewhere, or mostly copy/paste from other pre-existing code) should be `Copyright (c) 2020 Konduit K.K.`
There's other instances, I won't flag all of them.
Also missing copyright headers on EncodableToINDArrayTransform for example

**Review by AlexDBlack — CHANGES_REQUESTED** [review]
No objections to the datavec API changes. :+1:
Flagged some copyright header issues

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/observation/transform/FilterOperation.java (hunk 4 above)
Comment (saudet) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/observation/transform/FilterOperation.java]: You know, I'm not sure about the whole copyright thing. I think the copyright remains with the authors, and doesn't move to either Konduit or Eclipse from what I understand.

Comment (saudet) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/observation/transform/FilterOperation.java]: See here: https://www.eclipse.org/legal/legalfaq.php#h.r2y1yck33rg4

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/helper/INDArrayHelper.java (hunk 2 above)
Comment (aboulang2002) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/helper/INDArrayHelper.java:29]: Ok, I'll keep a note to check this later. Thanks

File: datavec/datavec-api/src/main/java/org/datavec/api/transform/Operation.java (hunk 3 above)
Comment (aboulang2002) [reply datavec/datavec-api/src/main/java/org/datavec/api/transform/Operation.java]: done

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/observation/transform/FilterOperation.java (hunk 4 above)
Comment (aboulang2002) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/observation/transform/FilterOperation.java]: I have changed those that had a header. There are still a couple of files without headers, but if I fix them, that will bring this PR beyond 1000 lines. I can add the missing headers in the part 2 of this contribution, if you don't mind.

**Review by saudet — APPROVED** [review]

**Review by AlexDBlack — APPROVED** [review]

**Pull request merged**